### PR TITLE
ICD manager Implemenation

### PR DESCRIPTION
**Pull request by jmartinez-silabs**
This PR implements the ICD manager using a timer-based and eventing mechanism. It uses the ICD cluster variables to determine the periods in which the ICD manager needs to be in Active or Idle mode.

Additionally, the Commissioning window and failsafe status can keep the ICD in Active Mode for a longer period than the standard Active mode interval specified in the ICD attributes.

Notes:

- This implementation currently only supports SIT ICDs. LIT support is not complete.
- most of the changes are gated by the `CHIP_CONFIG_ENABLE_ICD_SERVER` and does not impact any app not built as an ICD.
- The Silabs App can now use the ICD options. ICD cluster needs to be enabled on other apps as it is only enabled on our lighting-app. I will configure others in a subsequent PR
- The old SED options are currently maintained but will be phased out. I intend to remove it completely in the upcoming weeks.

### CODE REVIEW

File: src/app/icd/BUILD.gn
Comment (mkardous-silabs) [inline src/app/icd/BUILD.gn:37]: I think a batter location would be putting this config in the build.gn for tests.

File: src/app/icd/ICDEventManager.cpp
Comment (mkardous-silabs) [inline src/app/icd/ICDEventManager.cpp:29]: would it be better of having the init in the `server.cpp` file before passing it to the ICDEventManager.h?

File: src/app/icd/ICDManager.cpp
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:50]: can the feature map change at runtime? might be better if it as an attribute of the class instead.

File: src/app/icd/ICDManager.cpp
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:61]: The min polling interval is not a hard requirement. we don't need the check at this point.

File: src/app/icd/ICDManager.cpp
Comment (mkardous-silabs) [inline src/app/icd/ICDManager.cpp:89]: this could a VerifyOrReturn macro that does nothing, less lines.

File: src/messaging/ReliableMessageMgr.cpp
Comment (mkardous-silabs) [inline src/messaging/ReliableMessageMgr.cpp:258]:
```suggestion
    //   "An ICD sender SHOULD increase t to also account for its own sleepy interval
```

File: src/messaging/ReliableMessageProtocolConfig.cpp
Comment (mkardous-silabs) [inline src/messaging/ReliableMessageProtocolConfig.cpp:69]: I think we need to add a todo here to check if the device is a LIT ICD to not advertise the SII key.

**Review by jepenven-silabs — APPROVED** [review]

File: src/app/icd/ICDManager.cpp
Comment (jmartinez-silabs) [reply src/app/icd/ICDManager.cpp:50]: Featuremap is a read only attribute, It cannot change at run time if I am not mistaken.

File: src/app/icd/ICDManager.cpp
Comment (jmartinez-silabs) [reply src/app/icd/ICDManager.cpp:61]: good point I thought it was a SHALL. Maybe it should be a SHALL. since it is not active code yet I won't touch it but let's decide if it should be a SHALL or stay a should. 

Either way, we will need a way to differentiate a LIT from an SIT at run time and both can have the CIP

File: src/messaging/ExchangeContext.cpp
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:654]: This was posting the message received event.

File: src/messaging/ExchangeContext.cpp
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:650]: How do we know this was a message received?  It might have been a message sent.  Or a response wait timeout.

File: src/messaging/ExchangeContext.cpp
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:1]: A node typically determines whether it is in Active or Idle mode based on whether it has any outstanding Exchanges in the Message Layer. While there are Exchanges active, a node typically will remain in Active mode. Once all Exchanges are closed and the node has no other outstanding reasons for staying active, a node SHOULD transition into Idle mode. An ICD node MAY use additional modes and logic for switching between them.

But what I don't see is how that bit about "stay active while there is an outstanding exchange" is implemented in this new setup.  And the answer seems to be that it's not...

File: src/system/SystemLayerImplFreeRTOS.cpp
Comment (bzbarsky-apple) [inline src/system/SystemLayerImplFreeRTOS.cpp:84]: This doesn't make sense.  On the left we have a _timestamp_.  On the right we have a _duration_.

Was this tested?

File: src/system/SystemLayerImplSelect.cpp
Comment (bzbarsky-apple) [inline src/system/SystemLayerImplSelect.cpp:206]: What happens if the timer is in `mExpiredTimers` at this point?

File: src/system/SystemLayerImplSelect.cpp
Comment (bzbarsky-apple) [inline src/system/SystemLayerImplSelect.cpp:218]: Again, what if it's in mExpiredTimers?

File: src/system/SystemTimer.cpp
Comment (bzbarsky-apple) [inline src/system/SystemTimer.cpp:175]: Given this implementation, IsTimerActive will return false for timers that are not-yet-fired but whose time has expired while we were running this code.

I strongly suggest pulling the timer changes out into a separate PR that:

1) Will be able to get careful reviews.
2) Will have unit tests covering all the exciting edge cases here.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/BUILD.gn
Comment (jmartinez-silabs) [reply src/app/BUILD.gn:62]: Yes, it is not there to stay I should have put a to-do. Trying to fix/find why this define is causing issues in some test and not others. 

I'll try to get right of it either in this PR or else as soon a s possible

File: src/app/icd/ICDEventManager.cpp
Comment (jmartinez-silabs) [reply src/app/icd/ICDEventManager.cpp:59]: Yes, didn't end up using it so far. I can remove it for now

File: src/app/icd/ICDManager.cpp
Comment (jmartinez-silabs) [reply src/app/icd/ICDManager.cpp:84]: You are right i had a if return there instead and forgot to inverse the condition when cleaning up. Thanks

File: src/app/FailSafeContext.cpp
Comment (jmartinez-silabs) [reply src/app/FailSafeContext.cpp:60]: Yes I'm a subsequent pr I will remove all old SED usage and fix other platforms to use ICD

File: src/messaging/ExchangeContext.cpp
Comment (jmartinez-silabs) [reply src/messaging/ExchangeContext.cpp:1]: My assumption was that while message are being sent , the activeThreshold timer would keep the session active long enough for the next message to me sent, until all messages are sent. The spec changed recently about the activeThreshold minimal value. So I do need to take care of that now. It will be added to the ICD manager

File: src/system/SystemLayerImplFreeRTOS.cpp
Comment (jmartinez-silabs) [reply src/system/SystemLayerImplFreeRTOS.cpp:84]: The return type.of remainingTime is wrong. It the timestamp - the current timestamps. This code should be ok. I'll fix the type

File: src/system/SystemTimer.cpp
Comment (jmartinez-silabs) [reply src/system/SystemTimer.cpp:175]: In the case of the ICD usage. Timer expired will also allow me  to extend the active time by restarting the timer for an activeThreshold interval which. That work for my use case, but I guess in the general sense of the extendTimeraTo and IsActive it would be better to make a distinction.

File: src/app/BUILD.gn
Comment (bzbarsky-apple) [reply src/app/BUILD.gn:62]: Followup is fine as long as we document what's going on.

File: src/messaging/ExchangeContext.cpp
Comment (bzbarsky-apple) [reply src/messaging/ExchangeContext.cpp:1]: Here's the thing: let's say our ActiveThreshold timer is 5s or whatever it is.

We send a message.  We wait for the other side to process it.  This can easily be 6-30 seconds depending on the message and the MRP settings involved; in particular it can easily be longer than our ActiveThreshold period.

We _have_ to stay in active mode while we are waiting for that response, even if the response timeout is longer than ActiveThreshold.  ActiveThreshold kicks in once we have no more exchanges waiting for responses; an exchange waiting for a response _has_ to keep active mode going while it's waiting, because otherwise we can't receive the response.

File: src/system/SystemTimer.cpp
Comment (bzbarsky-apple) [reply src/system/SystemTimer.cpp:175]: >  So the new separate PR might get merged before.

That sounds perfect, right?

Comment (jmartinez-silabs) [reply src/system/SystemTimer.cpp:175]: Yes I agree

File: src/messaging/ExchangeContext.cpp
Comment (jmartinez-silabs) [reply src/messaging/ExchangeContext.cpp:650]: Thanks for clarifying the `MessageHandled`. I am moving this to `HandleMessage`. What is your opinion on this? Should we still notify the `ICDManager` for unexpected messages?

Comment (bzbarsky-apple) [reply src/messaging/ExchangeContext.cpp:650]: Realistically, I think this is generally the wrong model.  The right model is that we should be in active mode while an exchange is waiting for a response, no?  So the right notifications are:

1. We sent a message, waiting for a response (kicks us into active mode, where we need to stay until the next notification)
2. No longer waiting for a response (timeout or response received).  This is where we can release the "lock" on active mode, but stay in it until the timer is hit.

Comment (bzbarsky-apple) [reply src/messaging/ExchangeContext.cpp:650]: I guess if we send a message that does _not_ expect a response we should also do the threshold thing.

So if we want the ICDManager to implement the logic, the needed notifications are:

1. Message sent, no response expected.
2. Message sent, response expected; needs to include some way to identify _which message_ for purposes of notification 3.  Could be the message counter.
3. No longer waiting for a response for message (again, needs to include some way to indicate which message).

But I suspect in practice the correlation of messages and responses should just happen in the exchange itself, and the notifications should be:

1. Message sent, no response expected.
2. Message sent, waiting for a response.
3. No longer waiting for a response that we previously notified about.

File: src/messaging/ExchangeContext.cpp
Comment (jmartinez-silabs) [reply src/messaging/ExchangeContext.cpp:1]: I still need to implement this in a proper way.

File: src/messaging/ExchangeContext.cpp
Comment (jmartinez-silabs) [reply src/messaging/ExchangeContext.cpp]: Since there can be multiple ExchangeContext Sending the notification for 
- Message sent, waiting for a response.
- No longer waiting for a response that we previously notified about.
I need a way to identify/Match the notification that clears the wait. Or maybe a count of how many responses notifications I am expecting

Comment (bzbarsky-apple) [reply src/messaging/ExchangeContext.cpp]: Right, count is enough.  But you do need a count.

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.h
Comment (kkasperczyk-no) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.h:110]: It seems to me, that you want to split the SED and ICD cases in Thread platform. I wonder whether it's a good idea, as from the Thread platform perspective it should be transparent. No matter what heuristic is above, the Thread platform should care only about polling period updates. The rest depends on how the upper layers will use the platform API.

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp
Comment (jmartinez-silabs) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1968]: Everything tied with SED will just be deleted once ICD is fully implemented and used by platforms. I was keeping most of SED untouched for now to 
1) reduce the size of this PR
2) avoid breaking other platform while we lay the ground work for ICDs

I am just returning from vacation today, I am addressing the last comments today; it should be ready to be merged by tomorrow.

Once this is merged, with a few other pr (some are already merged) ICD SIT will be fully supported. The old SED code will then be removed.

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.h
Comment (jmartinez-silabs) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.h:110]: As mentioned in my other comment, ICD will replace all SED calls. All #ifdef related to SED will be removed and only ICD will be used.

In terms of OpenThread stack. Nothing changes, Thread doesn't care about how Matter manages an ICD. It will be the same at the low level, (be a sleepy MTD device, set your polling and/or csl timing). 
However now ICD will be in full control of when we are in Active mode or when we can go in idle.

File: src/messaging/ExchangeContext.cpp
Comment (jmartinez-silabs) [reply src/messaging/ExchangeContext.cpp]: Updated the eventing and "locking" in active mode as we discussed.

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.h
Comment (kkasperczyk-no) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.h:110]: Ok, great, my only concern was splitting ICD and SED functionalities. Once I'm sure it's only intermediate state before merging other PRs, I'm ok with that.

File: src/app/icd/ICDManager.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDManager.cpp:68]: This should presumably have a TODO comment pointing to an issue.

File: src/app/icd/ICDManager.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDManager.cpp:93]:
```suggestion
    // Active mode can be re-triggered.
```

File: src/app/icd/ICDManager.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDManager.cpp:100]: Check status?  Not sure what you should do on failure, though.

File: src/app/icd/ICDManager.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDManager.cpp:119]: Check status?

File: src/app/icd/ICDManager.cpp
Comment (bzbarsky-apple) [inline src/app/icd/ICDManager.cpp:131]: Check status.

File: src/include/platform/CHIPDeviceEvent.h
Comment (bzbarsky-apple) [inline src/include/platform/CHIPDeviceEvent.h:199]:
```suggestion
     * Signals that the commissioning window has opened or closed.
```

File: src/include/platform/CHIPDeviceEvent.h
Comment (bzbarsky-apple) [inline src/include/platform/CHIPDeviceEvent.h:262]:
```suggestion
     * This event contains a MessageSent structure.
```

File: src/include/platform/CHIPDeviceEvent.h
Comment (bzbarsky-apple) [inline src/include/platform/CHIPDeviceEvent.h:273]:
```suggestion
     * An Exchange Context that was waiting for a response is no longer waiting for it.
     *
     * This event can occur due to any of the following:
     *  - A response message was received.
     *  - An exchange context timed out waiting for a response.
     *  - The exchange context was closed while a response was expected.
     *
     * This event contains an RxEventContext structure.
```

File: src/include/platform/CHIPDeviceEvent.h
Comment (bzbarsky-apple) [inline src/include/platform/CHIPDeviceEvent.h:278]:
```suggestion
     * An application event occured that should wake up the system/device
```

File: src/include/platform/CHIPDeviceEvent.h
Comment (bzbarsky-apple) [inline src/include/platform/CHIPDeviceEvent.h:576]: Should document when this will be false.

File: src/messaging/ExchangeContext.cpp
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:242]: Should this just be "else" on the preceding if, if we change the condition like this?

File: src/messaging/ExchangeContext.cpp
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:247]: So this is going to kick us into active mode for a bit for sending:

1. Non-resend messages that expect a response.
2. Non-resend messages that don't expect a response.
3. Standalone acks.

But will not  kick us into active mode when we do an MRP retransmit of a message.

Is that the intended behavior?  If so, why is it the intended behavior?  I would really like to understand what part of the spec this is implementing and why we think the above description is the right implementation.

File: src/messaging/ExchangeContext.cpp
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:652]: So here we are going to kick things along in active mode on receipt of non-response messages, right?  Why?  Why do we not treat a duplicate or ack in the same way?  What is the spec basis for this behavior?

File: src/app/icd/ICDEventManager.cpp
Comment (jmartinez-silabs) [reply src/app/icd/ICDEventManager.cpp:25]: Isn't the Max exchange context relatively low? I see 16 by default and most platforms have it set to 8. I assumed at uint8_t holding up to 255 expected message count would be enough. I'll add the assert.

File: src/app/icd/ICDManager.cpp
Comment (jmartinez-silabs) [reply src/app/icd/ICDManager.cpp:68]: There is a to-do. I'll move it up at the start of the comments. Do you want me to link the todo to a git issue?

File: src/messaging/ExchangeContext.cpp
Comment (jmartinez-silabs) [reply src/messaging/ExchangeContext.cpp:246]: As discussed, I opened to a ticket for me to track and add the logic/event posting in the MRP retransmit.
https://github.com/project-chip/connectedhomeip/issues/27864

File: src/messaging/ExchangeContext.cpp
Comment (jmartinez-silabs) [reply src/messaging/ExchangeContext.cpp:651]: So based on our Slack conversation, I opened a spec Issue to add more details in the spec about the reasons that would require to be in active mode.
https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/7295

File: src/app/icd/ICDEventManager.cpp
Comment (bzbarsky-apple) [reply src/app/icd/ICDEventManager.cpp:25]: > Isn't the Max exchange context relatively low? 

It's configurable by applications, with platforms providing defaults.  If someone does configure it to be a few hundred, we don't want things to subtly break for them.

File: src/app/icd/ICDManager.cpp
Comment (bzbarsky-apple) [reply src/app/icd/ICDManager.cpp:72]: @jmartinez-silabs Yes, please link to issue here.

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by jmartinez-silabs** [comment]
Issues require a followup to this PR

https://github.com/project-chip/connectedhomeip/issues/27864
https://github.com/project-chip/connectedhomeip/issues/27875
https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/7295

**Pull request merged**
